src/CodeGen_GPU_Host.cpp

```diff
@@ -511,6 +527,47 @@ Value *CodeGen_GPU_Host<CodeGen_CPU>::get_module_state(const std::string &api_un
     return module_state;
 }
 
+template<typename CodeGen_CPU>
+void CodeGen_GPU_Host<CodeGen_CPU>::visit(const Call *op) {
+    CodeGen_CPU::visit(op);
+    if (op->name == "halide_device_malloc" || op->name == "halide_copy_to_device") {
+        // Register a destructor for this buffer if this is the first
+        // device_malloc or copy_to_device for it.
+        internal_assert(op->args.size() == 2);
+        const Variable *buf_var = op->args[0].as<Variable>();
+        internal_assert(buf_var);
+        const string &buf_name = buf_var->name;
+        string destructor_name = buf_name + "_destructor";
+
+        // We may already have a destructor for this allocation, if
+        // this is one of many copy_to_device calls.
+        if (!sym_exists(destructor_name)) {
+            llvm::Value *buf = sym_get(buf_name);
+            // Register a destructor that frees the device allocation.
+            llvm::Instruction *destructor =
+                register_destructor(module->getFunction("halide_device_free_as_destructor"), buf);
+            sym_push(destructor_name, destructor);
+        }
+    }
+}
+
+template<typename CodeGen_CPU>
+void CodeGen_GPU_Host<CodeGen_CPU>::visit(const Free *op) {
+    CodeGen_CPU::visit(op);
+
+    // Also free gpu memory by triggering the destructor
+    string destructor_name = op->name + ".buffer_destructor";
```

Comment:
Never mind, I see now, the Call directly references the .buffer member, correct?